### PR TITLE
Fixes queen not being able to spawn

### LINKED ISSUE #5999: Queen fails to respawn sometimes when selecting a location
### Testmerges

#5724, #5763 , #5829, #5006 , #5401 ,#5938 , #5974

### Round ID

21445

### Description of the bug

I was trying to go queen on Shivas Snowball and selected "Research Hab Interior" after a short amount of time (10-45 seconds) of pondering my hive location. 

When I selected it, it did not spawn me in as queen and instead simply closed the GUI, leaving my placeholder character in an unplayable state. No queen spawns at all and it needs admin intervention to rectify so this is **extremely disruptive** for the game flow.

This has been reported before on the discord so I suspect that the map may not have something to do with it.  

### What's the difference with what should have happened?

Queen should be spawned into the selected location instead of being put into a non playable state.

### How do we reproduce this bug?

1. Go Shivas
2. Wait 20-30 seconds before selecting
3. Select "Research Hab Interior"
...

### Issue Bingo

- [ ] Issue could be reproduced at least once
- [X] Issue happened in a recent (less than 7 days ago) round
- [ ] Couldn't find an existing issue about this (https://github.com/cmss13-devs/cmss13/issues)

### PR DESCRIPTION
**Pull request by vero5123**
# About the pull request
Fixes #5999 

# Explain why it's good for the game
bug fix

# Changelog
:cl:
fix: fixes queen getting stuck in the lobby if they take too long
/:cl:

### CODE REVIEW

File: code/game/gamemodes/cm_initialize.dm [hunk 1]
```diff
@@ -687,9 +687,9 @@ Additional game mode variables.
 	return TRUE
 
 /// Pick and setup a queen spawn from landmarks, then spawns the player there alongside any required setup
-/datum/game_mode/proc/pick_queen_spawn(datum/mind/ghost_mind, hivenumber = XENO_HIVE_NORMAL)
+/datum/game_mode/proc/pick_queen_spawn(mob/new_player/NP, hivenumber = XENO_HIVE_NORMAL)
```
Comment (Drulikar) [inline code/game/gamemodes/cm_initialize.dm:690]: What is passed here isn't a `/mob/new_player`. Its a `/mob/living/carbon/human`
image
Also https://github.com/cmss13-devs/cmss13/blob/master/.github/guides/STYLES.md#dont-use-abbreviations

```suggestion
/datum/game_mode/proc/pick_queen_spawn(mob/player, hivenumber = XENO_HIVE_NORMAL)
```

File: code/game/gamemodes/cm_initialize.dm [hunk 2]
```diff
@@ -687,9 +687,9 @@ Additional game mode variables.
...
 	RETURN_TYPE(/turf)
-
+	var/datum/mind/ghost_mind = NP.mind
```
Comment (Drulikar) [inline code/game/gamemodes/cm_initialize.dm:692]:
```suggestion
	var/datum/mind/ghost_mind = player.mind
```

File: code/game/gamemodes/cm_initialize.dm [hunk 3]
```diff
@@ -710,6 +710,8 @@ Additional game mode variables.
 		spawn_list_map[spawn_name] = T
 
 	var/selected_spawn = tgui_input_list(original, "Where do you want you and your hive to spawn?", "Queen Spawn", spawn_list_map, QUEEN_SPAWN_TIMEOUT, theme="hive_status")
+	if(hive.living_xeno_queen)
+		NP.send_to_lobby()
```
Comment (Drulikar) [inline code/game/gamemodes/cm_initialize.dm:714]:
```suggestion
		player.send_to_lobby()
```

File: code/game/gamemodes/colonialmarines/xenovsxeno.dm [hunk 4]
```diff
@@ -154,7 +154,7 @@
 		var/mob/living/carbon/xenomorph/larva/L = new(xeno_turf, null, hivenumber)
 		ghost_mind.transfer_to(L)
 
-/datum/game_mode/xenovs/pick_queen_spawn(datum/mind/ghost_mind, hivenumber = XENO_HIVE_NORMAL)
+/datum/game_mode/xenovs/pick_queen_spawn(mob/new_player/NP, hivenumber = XENO_HIVE_NORMAL)
```
Comment (Drulikar) [inline code/game/gamemodes/colonialmarines/xenovsxeno.dm:157]:
```suggestion
/datum/game_mode/xenovs/pick_queen_spawn(mob/player, hivenumber = XENO_HIVE_NORMAL)
```

File: code/game/jobs/job/antag/xeno/queen.dm [hunk 5]
```diff
@@ -10,7 +10,7 @@
 	return spawn_positions
 
 /datum/job/antag/xenos/queen/transform_to_xeno(mob/new_player/NP, hive_index)
```
Comment (Drulikar) [inline code/game/jobs/job/antag/xeno/queen.dm:12]: Arguments should follow parent's implementation:
image

```suggestion
/datum/job/antag/xenos/queen/transform_to_xeno(mob/living/carbon/human/human_to_transform, hive_index)
```

File: code/game/jobs/job/antag/xeno/queen.dm [hunk 6]
```diff
@@ -10,7 +10,7 @@
...
-	SSticker.mode.pick_queen_spawn(NP.mind, hive_index)
+	SSticker.mode.pick_queen_spawn(NP, hive_index)
```
Comment (Drulikar) [inline code/game/jobs/job/antag/xeno/queen.dm:13]:
```suggestion
	SSticker.mode.pick_queen_spawn(human_to_transform, hive_index)
```

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/game/jobs/job/antag/xeno/queen.dm (hunk 5 above)
Comment (vero5123) [reply code/game/jobs/job/antag/xeno/queen.dm]: going to git line blame the guy before me that set that arg.

File: code/game/gamemodes/cm_initialize.dm [hunk 7]
```diff
@@ -710,6 +710,8 @@ Additional game mode variables.
 		spawn_list_map[spawn_name] = T
 
 	var/selected_spawn = tgui_input_list(original, "Where do you want you and your hive to spawn?", "Queen Spawn", spawn_list_map, QUEEN_SPAWN_TIMEOUT, theme="hive_status")
+	if(hive.living_xeno_queen)
+		player.send_to_lobby()
```
Comment (Drulikar) [inline code/game/gamemodes/cm_initialize.dm:714]:
```suggestion
		to_chat(original, SPAN_XENOANNOUNCE("You have taken too long to pick a spawn location, a queen has already evolved before you."))
		player.send_to_lobby()
```

**Review by Drulikar — CHANGES_REQUESTED** [review]
So this code is functional, even in the situation where someone evolves before the queen picks a location. But you ought to also adjust QUEEN_SPAWN_TIMEOUT to be ~1 minute or so. It took approximately 1 minute and 20s for me as a larva spawn to evolve to queen.

File: code/game/gamemodes/cm_initialize.dm [hunk 8]
```diff
@@ -687,9 +687,9 @@ Additional game mode variables.
 	return TRUE
 
 /// Pick and setup a queen spawn from landmarks, then spawns the player there alongside any required setup
-/datum/game_mode/proc/pick_queen_spawn(datum/mind/ghost_mind, hivenumber = XENO_HIVE_NORMAL)
+/datum/game_mode/proc/pick_queen_spawn(mob/player, hivenumber = XENO_HIVE_NORMAL)
 	RETURN_TYPE(/turf)
-
+	var/datum/mind/ghost_mind = player.mind
 	var/mob/living/original = ghost_mind.current
```
Comment (Drulikar) [inline code/game/gamemodes/cm_initialize.dm:693]: `original` could also probably be refactored out now too and player used instead. But not a big deal.

**Comment by vero5123** [comment]
> So this code is functional, even in the situation where someone evolves before the queen picks a location. 

Surprised?

**Comment by Drulikar** [comment]
> > So this code is functional, even in the situation where someone evolves before the queen picks a location.
> 
> Surprised?

Not what I'm conveying. What I was conveying is this probably wouldn't even be necessary if the timer didn't allow people to evolve before queen is forced to pick a location. But I don't know how client disconnects might get handled so the code is fine to have that logic since it *should* kick them back to lobby even if disconnected - or at least tries to.

**Review by Drulikar — APPROVED** [review]

**Pull request merged**